src/TimeoutMigrationTool/Asp/AspTimeoutsSource.cs

```diff
@@ -0,0 +1,608 @@
+            return migrationTable;
+        }
+
+        async Task<CloudTable> GetEndpointTimeoutTable(string endpointName, CloudTableClient tableClient)
+        {
+            var endpointTimeoutTableName = $"{tablePrefix}{timeoutTableName}";
+            var endpointTimeoutTable = tableClient.GetTableReference(endpointTimeoutTableName);
+
+            if (!await endpointTimeoutTable.ExistsAsync())
+            {
+                throw new Exception(
+                    $"The timeout table '{endpointTimeoutTableName}' of the endpoint '{endpointName}' was not found.");
+            }
+
+            return endpointTimeoutTable;
+        }
+
+        public async Task<IReadOnlyList<TimeoutData>> ReadBatch(int batchNumber)
+        {
+            var tableClient = CreateCloudTableClient();
+            var blobServiceClient = new BlobServiceClient(connectionString);
+            var blobContainerClient = blobServiceClient.GetBlobContainerClient(containerName);
+
+            var migrationTable = GetMigrationTable(tableClient);
+
+            var query = new TableQuery<MigratedTimeoutDataEntity>()
+                .Where(CombineFilters(GenerateFilterCondition("PartitionKey", QueryComparisons.Equal, batchNumber.ToString(CultureInfo.InvariantCulture)),
+                    TableOperators.And,
+                    GenerateFilterConditionForInt("BatchState", QueryComparisons.Equal, 0)));
+
+            TableContinuationToken token = null;
+            var timeoutData = new List<TimeoutData>(batchSize);
+            CancellationToken queryCancellationToken = CancellationToken.None;
+            do
+            {
+                // will issue a table scan for now
+                var seg = await migrationTable.ExecuteQuerySegmentedAsync(
+                        query: query,
+                        token: token,
+                        requestOptions: null,
+                        operationContext: null,
+                        cancellationToken: queryCancellationToken)
+                    .ConfigureAwait(false);
+                token = seg.ContinuationToken;
+
+                await foreach (var data in MaterializeTimeoutData(blobContainerClient, seg.Results,
+                    queryCancellationToken))
+                {
+                    timeoutData.Add(data);
+                }
+
+            } while (token != null && !queryCancellationToken.IsCancellationRequested);
+
+            return timeoutData;
+        }
+
+        async IAsyncEnumerable<TimeoutData> MaterializeTimeoutData(
+            BlobContainerClient blobContainerClient,
+            IEnumerable<MigratedTimeoutDataEntity> timeoutDataEntities,
+            [EnumeratorCancellation] CancellationToken cancellationToken = default,
+            int maxConcurrentBlobDownloads = 200) // a single blob can handle up to 500 req/s
+        {
+            using var throttler = new SemaphoreSlim(maxConcurrentBlobDownloads);
```

Comment:
Is there a way to pair this with a time element? At the moment it looks like we're just hoping that by limiting it to 200 that it will be under the 500/s limit. Is that a safe assumption?

Reply:
I asked myself exactly the same question. Then I realized the state address represents a blob and the state address is unique per timeout. So effectively we could almost go unbounded as far as blob storage is concerned but then I felt this might be crazy in terms of bandwidth consumption etc so I just set it to something arbitrary in between. Technically we could work with `Task.Delay` continuation magic but then we would also have to somehow sort by the destination address and stuff which to me sounds hugely overkill. Happy to be convinced otherwise though